### PR TITLE
[OpenVINO]Fix memory leak upon exception throwing

### PR DESCRIPTION
**Pull request by mosdav**
- Make sure to free the `output_shape` vector even if `Output names mismatch between OpenVINO and ONNX` exception is thrown
- Piggy back this PR to remove un-needed call to `fstream` `close` method

### CODE REVIEW

**Review by snnn — APPROVED** [review]

File: onnxruntime/core/providers/openvino/backend_utils.cc [hunk 1]
```diff
@@ -34,7 +34,6 @@ bool IsDebugEnabled() {
 void DumpOnnxModelProto(const Provider_ModelProto& model_proto, std::string file_name) {
```
Comment (snnn) [inline onnxruntime/core/providers/openvino/backend_utils.cc:34]: This piece of code is very bad. Please read the notes of https://man7.org/linux/man-pages/man2/close.2.html

A careful programmer should always check the return value of close/fclose, otherwise you may suffer data loss without notice. And it actually happens very often. I saw a company loss millons of dollars because of this, they put a corrupted model online.  I know C++ programmers love RAII, but correctness should be the top thing. 

Tl;dr Please don't use C++ fstream, use the C functions.

Comment (mosdav) [reply onnxruntime/core/providers/openvino/backend_utils.cc:34]: Alright, I'd love to hear what makes you think I'm the owner of this code :) maybe it would be constructive finding the owner (via `git blame`) so you could communicate your valuable feedback to her

Comment (snnn) [reply onnxruntime/core/providers/openvino/backend_utils.cc:34]: All the code are co-own by the whole team. You may think is a big family, all the people are responsible for cleaning up the rooms. :-) For example, if you want to use a table, and you found it is dirty, then it would be great if you can help make it clean. Otherwise you'll be uncomfortable to sit at there right? And we'll be grateful for the help you provide. I don't like to trace back to the people who made it dirty and blame him/her.  That's my personal thought.

**Review by snnn — APPROVED** [review]
Will merge it first and leave the remaining improvements as TODO.
